sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -478,77 +512,399 @@ static struct codec_info codec_info_list[] = {
+		if (!cpu_name)
+			return -ENOMEM;
+
+		init_dai_link(dai_links, id + j, name, 1 - i, i, cpus, cpu_name,
+			      codec, link->num_adr, NULL, NULL);
+		set_codec_init_func(link, dai_links + id + j, 1 - i);
+		j++;
+	}
+
+	return j;
+}
+
+static inline int port_mask_to_num(int pmask)
+{
+	int num = 0;
+
+	for (num = 0; pmask; pmask >>= 1)
+		num += pmask & 1;
+
+	return num;
+}
+
+static int sof_card_dai_links_create(struct device *dev,
+				     struct snd_soc_acpi_mach *mach,
+				     struct snd_soc_card *card)
+{
+	struct snd_soc_dai_link_component *idisp_components;
+	struct snd_soc_dai_link_component *ssp_components;
+	struct snd_soc_acpi_mach_params *mach_params;
+	int ssp_num, sdw_num, hdmi_num = 0, dmic_num;
+	const struct snd_soc_acpi_link_adr *adr_link;
+	struct snd_soc_dai_link_component *cpus;
+	struct snd_soc_dai_link *links;
+	int num_links, ssp_mask;
+	char *name, *cpu_name;
+	int i, j, id = 0;
+	int val;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_COMMON_HDMI_CODEC)
+	hdmi_num = sof_rt711_rt1308_rt715_quirk & SOF_SDW_TGL_HDMI ?
+				SOF_TGL_HDMI_COUNT : SOF_PRE_TGL_HDMI_COUNT;
 #endif
-	{
-		.name = "SDW2-Playback",
-		.id = 3,
-		.init = second_spk_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw2_pin2, sdw2_codec, platform),
-	},
-};
+
+	ssp_mask = SOF_SSP_GET_PORT(sof_rt711_rt1308_rt715_quirk);
+	ssp_num = port_mask_to_num(ssp_mask);
+
+	mach_params = &mach->mach_params;
+	sdw_num = get_sdw_dai_link_num(mach_params);
+
+	/* enable dmic01 & dmic16k */
+	dmic_num = sof_rt711_rt1308_rt715_quirk & SOF_SDW_PCH_DMIC ? 2 : 0;
+
+	dev_dbg(dev, "sdw %d, ssp %d, dmic %d, hdmi %d", sdw_num, ssp_num,
+		dmic_num, hdmi_num);
+
+	num_links = ssp_num + sdw_num + dmic_num + hdmi_num;
+	links = devm_kcalloc(dev, num_links, sizeof(struct snd_soc_dai_link),
+			     GFP_KERNEL);
+	cpus = devm_kcalloc(dev, num_links,
+			    sizeof(struct snd_soc_dai_link_component),
+			    GFP_KERNEL);
+	if (!links || !cpus)
+		return -ENOMEM;
+
+	/* SDW */
+	if (!sdw_num)
+		goto SSP;
+
+	adr_link = mach_params->links;
+	if (!adr_link)
+		return -EINVAL;
+
+	while (adr_link->num_adr) {
+		val = create_sdw_codec_dai(dev, links, id, cpus, adr_link);
+		if (val < 0) {
+			dev_err(dev, "failed to create dai link %d", id);
+			return -ENOMEM;
+		}
+
+		id += val;
+		adr_link++;
+	}
+
+SSP:
+	/* SSP */
+	if (!ssp_num)
+		goto DMIC;
+
+	for (i = 0, j = 0; ssp_mask; i++, ssp_mask >>= 1) {
+		struct codec_info *info;
+		char *codec_name;
+		int index;
+
+		if (!(ssp_mask & 0x1))
+			continue;
+
+		name = devm_kasprintf(dev, GFP_KERNEL,
+				      "SSP%d-Codec", i);
+		if (!name)
+			return -ENOMEM;
+
+		cpu_name = devm_kasprintf(dev, GFP_KERNEL, "SSP%d Pin", i);
+		if (!cpu_name)
+			return -ENOMEM;
+
+		val = sizeof(struct snd_soc_dai_link_component);
+		ssp_components = devm_kzalloc(dev, val, GFP_KERNEL);
+				if (!ssp_components)
+					return -ENOMEM;
+
+		index = find_codec_info(0, mach->id);
```

Comment:
does this really need to be done inside the loop?

Reply:
let's move it out of loop, thanks!